stan/math/rev/functor/algebra_solver_powell.hpp

```diff
@@ -21,57 +22,77 @@ namespace math {
  * function theorem. The call to Jacobian() occurs outside the call to
  * chain() -- this prevents malloc issues.
  */
-template <typename Fs, typename F, typename T, typename Fx>
+template <typename Fy, typename T, typename Fx>
 struct algebra_solver_vari : public vari {
-  /** vector of parameters */
-  vari** y_;
   /** number of parameters */
-  int y_size_;
+  const int y_size_;
+  /** value of parameters */
+  const Eigen::Matrix<T, Eigen::Dynamic, 1> y_val_;
+  /** System functor (f_) w.r.t. inputs (y_) */
+  Fy fy_;
+  /** array of parameters */
+  vari** y_;
```

Comment:
If you save all of `y` in an `arena_t` you won't need this (can save the vars themselves, and vars are just pointers to varis so that is equivalent to this).

Reply:
That sounds good, expect I'm not really sure how to rewrite this line without `fy_`:
https://github.com/stan-dev/math/blob/fe45cbb554effdae99e5184560873c430498ead7/stan/math/rev/functor/algebra_solver_powell.hpp#L91
I guess we could just construct the functor object _in_ the `chain()` method?